components/lib/state/src/main/java/mozilla/components/lib/state/Store.kt

```diff
@@ -0,0 +1,108 @@
+import java.util.concurrent.Executors
+
+typealias Reducer<S, A> = (S, A) -> S
+typealias Observer<S> = (S) -> Unit
+
+/**
+ * A generic store holding an immutable [State].
+ *
+ * The [State] can only be modified by dispatching [Action]s which will create a new state and notify all registered
+ * [Observer]s.
+ *
+ * @param initialState The initial state until a dispatched [Action] creates a new state.
+ * @param reducer A function that gets the current [State] and [Action] passed in and will return a new [State].
+ */
+open class Store<S : State, A : Action>(
+    initialState: S,
+    private val reducer: Reducer<S, A>
+) {
+    private val storeContext = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+    private val storeScope = CoroutineScope(storeContext)
+    private val subscriptions: MutableList<Subscription<S, A>> = mutableListOf()
+    private var currentState = initialState
+
+    /**
+     * The current [State].
+     */
+    val state: S
+        get() = currentState
+
+    /**
+     * Registers an [Observer] function that will be invoked whenever the [State] changes.
+     *
+     * @param receiveInitialState If true the observer function will be invoked immediately with the current state.
+     * @return A [Subscription] object that can be used to unsubscribe from further state changes.
+     */
+    fun observe(receiveInitialState: Boolean = true, observer: Observer<S>): Subscription<S, A> {
+        val subscription = Subscription(observer, store = this)
+
+        synchronized(subscriptions) {
+            subscriptions.add(subscription)
+        }
+
+        if (receiveInitialState) {
+            observer.invoke(currentState)
+        }
+
+        return subscription
+    }
+
+    /**
+     * Dispatch an [Action] to the store in order to trigger a [State] change.
+     */
+    fun dispatch(action: A) = storeScope.launch(storeContext) {
+        dispatchInternal(action)
+    }
+
+    @Synchronized
+    private fun dispatchInternal(action: A) {
+        val newState = reducer(currentState, action)
+
+        if (newState == currentState) {
+            // Nothing has changed.
+            return
+        }
+
+        currentState = newState
+
+        synchronized(subscriptions) {
+            subscriptions.forEach { it.observer.invoke(currentState) }
+        }
+    }
+
+    private fun removeSubscription(subscription: Subscription<S, A>) {
+        synchronized(subscriptions) {
+            subscriptions.remove(subscription)
+        }
+    }
+
+    /**
+     * A [Subscription] is returned whenever an observer is registered via the [observe] method. Calling [unsubscribe]
+     * on the [Subscription] will unregister the observer.
+     */
+    class Subscription<S : State, A : Action> internal constructor(
+        internal val observer: Observer<S>,
+        private val store: Store<S, A>
+    ) {
+        var binding: Binding? = null
+
+        fun unsubscribe() {
+            store.removeSubscription(this)
```

Comment:
This may be a no-op, if the store doesn't have this `subscription` anymore, right? Currently the only way to remove a subscription from the list is via this method, but perhaps that should either be documented, or somehow enforced?